flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -67,13 +77,36 @@ class Executing extends StateWithExecutionGraph implements ResourceListener {
         this.context = context;
         Preconditions.checkState(
                 executionGraph.getState() == JobStatus.RUNNING, "Assuming running execution graph");
+        this.scalingIntervalMin = scalingIntervalMin;
+        this.scalingIntervalMax = scalingIntervalMax;
+        this.lastRescale =
+                Instant.now(); // Executing is recreated with each restart (when we rescale)
+        // we consider the first execution of the pipeline as a rescale event
+        Preconditions.checkState(
+                !scalingIntervalMin.isNegative(),
+                "%s must be positive integer or 0",
+                JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MIN.key());
+        if (scalingIntervalMax != null) {
+            Preconditions.checkState(
+                    scalingIntervalMax.compareTo(scalingIntervalMin) > 0,
+                    "%s(%d) must be greater than %s(%d)",
+                    JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX.key(),
+                    scalingIntervalMax,
+                    JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MIN.key(),
+                    scalingIntervalMin);
+        }
 
         deploy();
 
         // check if new resources have come available in the meantime
         context.runIfState(this, this::maybeRescale, Duration.ZERO);
     }
 
+    @VisibleForTesting
+    void setLastRescale(Instant lastRescale) {
```

Comment:
Can the `lastRescale` be passed from constructor? 

If so, the `lastRescale` can be final, it's safer. I see it can works, please help check, thanks~

Reply:
No because `lastRescale` is never passed from the outside of the Executing object in production code: it is only initialized at the creation of Executing. `setLastRescale` is only used in test code. Changing a production constructor just for a test case is a bad design.

Reply:
> Changing a production constructor just for a test case is a bad design.

So, do you think change a filed from `final` to `non-final` is a good design? If the lastRescale is passed from constructor, the `lastRescale` can be final, it's safe and reasonable. The caller of `new Executing()` can call the  `Instant.now()`.

It is worth mentioning that this suggestion is not created by me, but is a standard in the Java industry. It is called by everyone: `Inject dependencies`. `Inject dependencies` is also mentioned in Flink's "Code Style and Quality Guide", which is in the chapter on `Design for Testability`[1]. Two of them describe the current code:

- Reusability becomes easier if constructors don’t create their dependencies (the objects assigned to the fields), but accept them as parameters.
- Effectively, constructors should have no new keyword.

So, I prefer the `lastRescale` is passed the the `Executing` constructor instead of new it in the `Executing` constructor.

[1] https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#design-for-testability

Reply:
Yes I know this and I agree but the big difference to me is that this field is a **constant** in the production code. But fair enough, I'll do the change if you want to.

Reply:
done